test/jdk/java/security/MessageDigest/TestCloneable.java

```diff
@@ -71,6 +75,50 @@ public void test(String algo, String provName)
             System.out.println(": NOT Cloneable");
             Assert.assertThrows(CNSE, ()->md.clone());
         }
+
+        System.out.print("Testing " + algo + " impl from " + provName);
+        final var d1 = MessageDigest.getInstance(algo, provName);
+        final var buffer = ByteBuffer.allocateDirect(1024);
+        final var r = new Random(1024);
+
+        fillBuffer(r, buffer);
+        d1.update(buffer); // this statement triggers tempArray allocation
+        final var d2 = (MessageDigest) d1.clone();
+        assert Arrays.equals(d1.digest(), d2.digest());
+
+        final var t1 = updateThread(d1);
+        final var t2 = updateThread(d2);
+        t1.join();
+        t2.join();
+
+        System.out.println(": Shared data check");
+        if (!Arrays.equals(d1.digest(), d2.digest())) {
```

Comment:
In the case where ByteBuffer has allocated a direct buffer, a new internal byte array gets allocated. This is the shared memory that's causing the problem. See MessageDigestSpi.java line 220.

Oops got to update the copyright.

Reply:
I think you mean line 126? :

`                tempArray = new byte[n];  `

Reply:
You are correct. How did I come up with 220?

Reply:
@mpdonova I believe it is because Random is producing the same sequence of bytes for each thread, and thus each MessageDigest should be equal. When the memory is shared, they inevitably overwrite each other's tempArray and you get different results.

Reply:
@mcpowers Some comments in the code explaining this better would help.

Reply:
That makes sense, thanks @seanjmullan

Reply:
@mcpowers can you also add a comment to the test, maybe right before line 95, with something like I said above: 

"Random is producing the same sequence of bytes for each thread, and thus each MessageDigest should be equal. When the memory is shared, they inevitably overwrite each other's tempArray and you get different results."

Reply:
@mcpowers can you also add a comment to the test, maybe right before line 95, with something like I said above: 

"Random is producing the same sequence of bytes for each thread, and thus each MessageDigest should be equal. When the memory is shared, they inevitably overwrite each other's tempArray and you get different results."

Reply:
Good idea. It's not obvious what's going on.